File: scaffold/src/main/java/top/xuqingquan/web/HttpHeaders.java
package top.xuqingquan.web;

import android.text.TextUtils;
import androidx.annotation.NonNull;
import androidx.collection.ArrayMap;

import java.util.Map;

public class HttpHeaders {
	public static HttpHeaders create() {
		return new HttpHeaders();
	}

	private final Map<String, Map<String, String>> mHeaders;

	HttpHeaders() {
		mHeaders = new ArrayMap<>();
	}

	public Map<String, String> getHeaders(String url) {
		String subUrl = subBaseUrl(url);
		if (mHeaders.get(subUrl) == null) {
			Map<String, String> headers = new ArrayMap<>();
			mHeaders.put(subUrl, headers);
			return headers;
		}
		return mHeaders.get(subUrl);
	}

	public void additionalHttpHeader(String url, String k, String v) {
		if (null == url) {
			return;
		}
		url = subBaseUrl(url);
		Map<String, Map<String, String>> mHeaders = getHeaders();
		Map<String, String> headersMap = mHeaders.get(subBaseUrl(url));
		if (null == headersMap) {
			headersMap = new ArrayMap<>();
		}
		headersMap.put(k, v);
		mHeaders.put(url, headersMap);
	}


	public void additionalHttpHeaders(String url, Map<String, String> headers) {
		if (null == url) {
			return;
		}
		String subUrl = subBaseUrl(url);
		Map<String, Map<String, String>> mHeaders = getHeaders();
		Map<String, String> headersMap = headers;
		if (null == headersMap) {
			headersMap = new ArrayMap<>();
		}
		mHeaders.put(subUrl, headersMap);
	}

	public void removeHttpHeader(String url, String k) {
		if (null == url) {
			return;
		}
		String subUrl = subBaseUrl(url);
		Map<String, Map<String, String>> mHeaders = getHeaders();
		Map<String, String> headersMap = mHeaders.get(subUrl);
		if (null != headersMap) {
			headersMap.remove(k);
		}
	}

	public boolean isEmptyHeaders(String url) {
		url = subBaseUrl(url);
		Map<String, String> heads = getHeaders(url);
		return heads == null || heads.isEmpty();
	}

	public Map<String, Map<String, String>> getHeaders() {
		return this.mHeaders;
	}

	private String subBaseUrl(String originUrl) {
		if (TextUtils.isEmpty(originUrl)) {
			return originUrl;
		}
		int index = originUrl.indexOf("?");
		if (index <= 0) {
			return originUrl;
		}
		return originUrl.substring(0, index);
	}

	@Override
	@NonNull
	public String toString() {
		return "HttpHeaders{" +
				"mHeaders=" + mHeaders +
				'}';
	}
}

File: scaffold/src/main/java/top/xuqingquan/web/agent/HttpHeaders.kt
package top.xuqingquan.web.agent

import android.text.TextUtils
import androidx.collection.ArrayMap

class HttpHeaders internal constructor() {

    val headers = ArrayMap<String, MutableMap<String, String>>()

    fun getHeaders(url: String?): MutableMap<String, String>? {
        val subUrl = subBaseUrl(url)
        if (headers[subUrl] == null) {
            val headers = ArrayMap<String, String>()
            this.headers[subUrl!!] = headers
            return headers
        }
        return headers[subUrl]
    }

    fun additionalHttpHeader(o_url: String?, k: String, v: String) {
        var url: String? = o_url ?: return
        url = subBaseUrl(url)
        val mHeaders = this.headers
        var headersMap: MutableMap<String, String>? = mHeaders[subBaseUrl(url)]
        if (null == headersMap) {
            headersMap = ArrayMap()
        }
        headersMap[k] = v
        mHeaders[url!!] = headersMap
    }


    fun additionalHttpHeaders(url: String?, headers: MutableMap<String, String>) {
        url ?: return
        val subUrl = subBaseUrl(url)
        val mHeaders = this.headers
        var headersMap: MutableMap<String, String>? = headers
        if (null == headersMap) {
            headersMap = ArrayMap()
        }
        mHeaders[subUrl!!] = headersMap
    }

    fun removeHttpHeader(url: String?, k: String) {
        url ?: return
        val subUrl = subBaseUrl(url)
        val mHeaders = headers
        val headersMap = mHeaders[subUrl]
        headersMap?.remove(k)
    }

    fun isEmptyHeaders(o_url: String?): Boolean {
        var url = o_url
        url = subBaseUrl(url)
        val heads = getHeaders(url)
        return heads.isNullOrEmpty()
    }

    private fun subBaseUrl(originUrl: String?): String? {
        if (TextUtils.isEmpty(originUrl)) {
            return originUrl
        }
        val index = originUrl!!.indexOf("?")
        return if (index <= 0) {
            originUrl
        } else {
            originUrl.substring(0, index)
        }
    }

    companion object {
        @JvmStatic
        fun create(): HttpHeaders {
            return HttpHeaders()
        }
    }
}
